src/transformers/configuration_utils.py

```diff
@@ -270,6 +274,14 @@ def __init__(self, **kwargs):
         else:
             self.num_labels = kwargs.pop("num_labels", 2)
 
+        if self.torch_dtype is not None and isinstance(self.torch_dtype, str):
+            # we will start using self.torch_dtype in v5, but to be consistent with
+            # from_pretrained's torch_dtype arg convert it to an actual torch.dtype object
+            if is_torch_available():
+                import torch
+
+                self.torch_dtype = getattr(torch, self.torch_dtype)
```

Comment:
OK, so torch is not always available when `config.torch_dtype is not None` - so now `config.torch_dtype` isn't always of `torch.dtype`.

I'm thinking perhaps this whole approach needs to be redone and only use the "float32", "float16", etc. strings everywhere, including `from_pretrained` and `from_config` args. And only convert to `torch.dtype` at the point it's used.

That way `torch_dtype` doesn't need to have a special handling at config level.

I hope this is recent/experimental enough that it's ok that we break the API.

Actually, if we do that, why even bother with `torch_` in `torch_dtype` and not just rename it to `dtype` - perhaps non-pt frameworks could tap into it as well? After all fp16-saved data by torch isn't any different from flux or tf, no?

Reply:
I can't think of a scenario where one would want one dtype for one framework and another dtype for another - so changing it to `dtype` sounds good to me.

Reply:
So the main concern is the back-compat in the API arg `torch_dtype` - if it's OK if we break it, then I propose both the config and the arg in `from_pretrained` and `from_config` to be just a `dtype` as a string: . "float32", "float16", etc. 

And then in the case of torch we convert it to the right `torch.dtype` on the fly.

Sylvain is not here for another week. Do you both support this breaking API change, @LysandreJik and @patrickvonplaten?

So instead of:
```
            torch_dtype (:obj:`str` or :obj:`torch.dtype`, `optional`):
                Override the default ``torch.dtype`` and load the model under this dtype. If ``"auto"`` is passed the
                dtype will be automatically derived from the model's weights.
```

It will be:
```
            torch_dtype (:obj:`str`, `optional`):
                Override the default ``dtype`` and load the model under this dtype, (e.g., ``"float16"``). If ``"auto"`` is passed the
                dtype will be automatically derived from the model's weights.
```

Reply:
On the other hand we already have `dtype` attribute in modeling_utils, which points to `torch.dtype`

https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_utils.py#L205

So my proposal might be confusing.

Should I call it `dtype_str` perhaps?

Reply:
If no version with `torch_dtype` has been released yet, I'm fine with changing it to `dtype`. However, note that in Flax we already have a `dtype` variable that is used however to define the precision the matmul operations are run in instead of the precision of the actual weights. In Flax we would like to take this design: https://github.com/huggingface/transformers/pull/13098as outlined by @patil-suraj

Reply:
There was a release since the feature was added.

Reading the PR you linked, I assume you're referring to a suggestion to disambiguate `dtype` with `params_dtype`, correct?

Reply:
we can also merge this as it fixes a bug, and then do another iteration to improve things, perhaps once Sylvain is back?

Please feel free to merge this if you'd like it asap.

Reply:
I would feel more comfortable to merge this PR without breaking changes and keeping the parameter called `torch_dtype`. Think it's worth to have a separate discussion here regarding a framework-agnostic `dtype` parameter for all PyTorch, Tensorflow, and Flax once @sgugger is back. @LysandreJik - feel free to merge if you agree!

Reply:
@stas00 regarding the PR: https://github.com/huggingface/transformers/pull/13098 - the idea of the PR is exactly to disentangle parameter dtype from matmul/computation dtype. In Flax, it's common practice that the `dtype` parameter defines the matmul/computation dtype, see: https://flax.readthedocs.io/en/latest/_autosummary/flax.linen.Dense.html#flax.linen.Dense.dtype instead of the parameter dtype and not the parameter dtype. 

So for Flax, I don't really think it would make sense to use a `config.dtype` to define weights dtype as it would be quite confusing with Flax's computation dtype parameter.

Reply:
> I would feel more comfortable to merge this PR without breaking changes and keeping the parameter called `torch_dtype`. 

Works for me. Although this API is bad at the moment due to inconsistent values in the config file and the function - the former a string, the latter a `torch.dtype`. Perhaps I can change these to support both string "float32" and torch.dtype in the same param of the function.

> Think it's worth to have a separate discussion here regarding a framework-agnostic `dtype` parameter for all PyTorch, Tensorflow, and Flax once @sgugger is back. 

Agreed!

Reply:
> In Flax, it's common practice that the dtype parameter defines the matmul/computation dtype, see: https://flax.readthedocs.io/en/latest/_autosummary/flax.linen.Dense.html#flax.linen.Dense.dtype instead of the parameter dtype and not the parameter dtype.

So it's somewhat similar to `dtype` arg in the new pytorch `autocast` feature it seems then, correct? (before it was a hardcoded fp16, but now it has a `dtype` arg to support bf16 too.

Reply:
Sounds good to me to keep `torch_dtype` and have a separate discussion for the framework-agnostic `dtype` parameter, to which `torch_dtype` could be an alias to prevent breaking changes to the existing API.

Reply:
Sounds like a plan, @LysandreJik 

Issue created: https://github.com/huggingface/transformers/issues/13246